src/libPMacc/include/Environment.hpp

```diff
@@ -38,160 +38,174 @@
+    void initGrids(DataSpace<DIM> gridSizeGlobal, DataSpace<DIM> gridSizeLocal, DataSpace<DIM> gridOffset)
+    {
+        PMacc::SubGrid<DIM>::getInstance().init(gridSizeLocal, gridSizeGlobal, gridOffset);
 
-        void initGrids(DataSpace<DIM> gridSizeGlobal, DataSpace<DIM> gridSizeLocal, DataSpace<DIM> gridOffset)
-        {
-            PMacc::SubGrid<DIM>::getInstance().init(gridSizeLocal, gridSizeGlobal, gridOffset);
+        EnvironmentController::getInstance();
 
-            EnvironmentController::getInstance();
+        DataConnector::getInstance();
 
-            DataConnector::getInstance();
+        PluginConnector::getInstance();
 
-            PluginConnector::getInstance();
+        nvidia::memory::MemoryInfo::getInstance();
 
-            nvidia::memory::MemoryInfo::getInstance();
+    }
 
-        }
+    void finalize()
+    {
+    }
 
-        void finalize()
-        {
-        }
+private:
+
+    Environment()
+    {
+    }
+
+    Environment(const Environment&);
 
-    private:
+    Environment& operator=(const Environment&);
 
-        Environment()
+    void setDevice(int deviceNumber)
+    {
+        int num_gpus = 0; //number of gpus
+        cudaGetDeviceCount(&num_gpus);
+        //##ERROR handling
+        if (num_gpus < 1) //check if cuda device is found
+        {
+            throw std::runtime_error("no CUDA capable devices detected");
+        }
+        else if (num_gpus < deviceNumber) //check if device can be selected by deviceNumber
         {
+            std::cerr << "no CUDA device " << deviceNumber << ", only " << num_gpus << " devices found" << std::endl;
+            throw std::runtime_error("CUDA capable devices can't be selected");
         }
 
-        Environment(const Environment&);
 
-        Environment& operator=(const Environment&);
+        int maxTrys = num_gpus;
 
-        void setDevice(int deviceNumber)
+        cudaDeviceProp devProp;
+        cudaError rc;
+        CUDA_CHECK(cudaGetDeviceProperties(&devProp, deviceNumber));
+        if (devProp.computeMode == cudaComputeModeDefault)
         {
-            int num_gpus = 0; //number of gpus
-            cudaGetDeviceCount(&num_gpus);
-            //##ERROR handling
-            if (num_gpus < 1) //check if cuda device is found
-            {
-                throw std::runtime_error("no CUDA capable devices detected");
-            }
-            else if (num_gpus < deviceNumber) //check if device can be selected by deviceNumber
+            /*if the gpu compute mode is set to default we use the given `deviceNumber`*/
+            maxTrys = 1;
+        }
+
+        for (int deviceOffset = 0; deviceOffset < maxTrys; ++deviceOffset)
+        {
+            const int tryDeviceId = (deviceOffset + deviceNumber) % num_gpus;
+            rc = cudaSetDevice(tryDeviceId);
+            if (rc == cudaSuccess)
             {
-                std::cerr << "no CUDA device " << deviceNumber << ", only " << num_gpus << " devices found" << std::endl;
```

Comment:
I don't even understand why a `cudaSuccess` resulted in a `std::runtime_error` in the *old* code ... o.0